homeassistant/components/nordpool/__init__.py

```diff
@@ -25,24 +29,52 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     return True
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: NordPoolConfigEntry) -> bool:
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: NordPoolConfigEntry
+) -> bool:
     """Set up Nord Pool from a config entry."""
 
-    coordinator = NordPoolDataUpdateCoordinator(hass, entry)
+    await cleanup_device(hass, config_entry)
+
+    coordinator = NordPoolDataUpdateCoordinator(hass, config_entry)
     await coordinator.fetch_data(dt_util.utcnow())
     if not coordinator.last_update_success:
         raise ConfigEntryNotReady(
             translation_domain=DOMAIN,
             translation_key="initial_update_failed",
             translation_placeholders={"error": str(coordinator.last_exception)},
         )
-    entry.runtime_data = coordinator
+    config_entry.runtime_data = coordinator
 
-    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    await hass.config_entries.async_forward_entry_setups(config_entry, PLATFORMS)
 
     return True
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: NordPoolConfigEntry) -> bool:
+async def async_unload_entry(
+    hass: HomeAssistant, config_entry: NordPoolConfigEntry
+) -> bool:
     """Unload Nord Pool config entry."""
-    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    return await hass.config_entries.async_unload_platforms(config_entry, PLATFORMS)
+
+
+async def cleanup_device(
+    hass: HomeAssistant, config_entry: NordPoolConfigEntry
+) -> None:
+    """Cleanup device and entities."""
+    device_reg = dr.async_get(hass)
+    entity_reg = er.async_get(hass)
+
+    entries = dr.async_entries_for_config_entry(device_reg, config_entry.entry_id)
+    for area in config_entry.data[CONF_AREAS]:
+        for entry in entries:
+            if entry.identifiers == {(DOMAIN, area)}:
+                continue
+
+            LOGGER.debug("Removing device %s", entry.name)
+            entities = er.async_entries_for_device(entity_reg, entry.id)
+            for _entity in entities:
+                entity_reg.async_remove(_entity.entity_id)
```

Comment:
```suggestion
```
entities will be cleaned up automatically once the config entry is removed from the device. The below code is preferred because other integrations might be connected  to the same device

Reply:
> The below code is preferred because other integrations might be connected to the same device.

The other method uses `async_update_device` but I guess there is some reason why it exists, but I'll take a look at that some other time.